src/TestFramework/Coverage/BufferedSourceFileFilter.php

```diff
@@ -0,0 +1,115 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\TestFramework\Coverage;
+
+use function array_key_exists;
+use Infection\FileSystem\FileFilter;
+use Infection\FileSystem\SourceFileFilter;
+use function Pipeline\take;
+use Symfony\Component\Finder\SplFileInfo;
+use Webmozart\Assert\Assert;
+
+/**
+ * We need to remove traces that are not in the list of source files,
+ * which could have files that were been directly specified. All the
+ * while later we may need a list of files that in the list but were
+ * not covered.
+ *
+ * On the other hand we don't need to filter traces all over again as
+ * we're checking them against pre-filtered list of files.
+ *
+ * @internal
+ */
+final class BufferedSourceFileFilter implements FileFilter
+{
+    private $filter;
+
+    /**
+     * An associative array mapping real paths to SplFileInfo objects.
+     *
+     * @var array<string, SplFileInfo>
+     */
+    private $sourceFiles = [];
+
+    /**
+     * @param SourceFileFilter|FileFilter $filter
+     * @param iterable<SplFileInfo> $sourceFiles
+     */
+    public function __construct(
+        FileFilter $filter,
+        iterable $sourceFiles
+    ) {
+        $this->filter = $filter;
+
+        // Make a list of source files so we can check against it.
+        foreach ($this->filter->filter(
+            $sourceFiles
+        ) as $filteredSourceFile) {
+            /** @var SplFileInfo $filteredSourceFile */
+            $sourceFilePath = $filteredSourceFile->getRealPath();
+
+            Assert::string($sourceFilePath);
+
+            $this->sourceFiles[$sourceFilePath] = $filteredSourceFile;
+        }
+    }
+
+    public function filter(iterable $input): iterable
+    {
+        return take($this->filter->filter($input))
+            ->filter(function (Trace $trace) {
+                $traceRealPath = $trace->getSourceFileInfo()->getRealPath();
+
+                Assert::string($traceRealPath);
+
+                if (array_key_exists($traceRealPath, $this->sourceFiles)) {
+                    unset($this->sourceFiles[$traceRealPath]);
+
+                    return true;
+                }
+
+                return false;
+            });
+    }
+
+    /**
+     * @return SplFileInfo[]
+     */
+    public function getUnseenFiles(): array
```

Comment:
only after debugging with xdebug I understood what this is used for. What do you think about renaming it to for example `unseenInCoverageReportFiles`.

As far as I understood, this method returns files that **are** in `source.directories` from `infection.json.dist` but **not** in coverage report (phpunit's `filter.whitelist.directory`)

Reply:
That's exactly what it is and I will rename it to match.